### PR TITLE
Stop test execution after first failure

### PR DESCRIPTION
**Pull request by acourtneybrown**
### Context
Addresses #1310

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/processors/MaxNParallelTestClassProcessor.java [hunk 1]
```diff
@@ -76,9 +83,29 @@ public void processTestClass(TestClassRunInfo testClass) {
...
+                processorsLock.unlock();
+            }
+            stoppable.add(actors).add(resultProcessorActor).stop();
         } catch (DispatchException e) {
             throw UncheckedException.throwAsUncheckedException(e.getCause());
         }
     }
+
+    @Override
+    public void stopNow() {
```
Comment (acourtneybrown) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/processors/MaxNParallelTestClassProcessor.java]: After more thought & local experiments, I realized that this method isn't going to work as anticipated because the `stop()` will likely have been invoked, clearing out the `processors` list & invalidating the associated `Actor`s.  I'm now thinking that this class needs to retain the "raw" `TestClassProcessor`s so that they can have `stopNow` called directly without going through the (potentially stopped) `Actor`.

Comment (acourtneybrown) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/processors/MaxNParallelTestClassProcessor.java]: This should be better/fixed with the [next commit](https://github.com/gradle/gradle/pull/4190/commits/e7607102bec0ffadcd3770ef5133c0b0e4e07145)

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 2]
```diff
@@ -544,6 +549,12 @@ public AbstractTestTask setTestNameIncludePatterns(List<String> testNamePattern)
         return this;
     }
 
+    @Option(option = "fail-fast", description = "Stops test execution after the first failed test")
+    public AbstractTestTask setFailFast(boolean failFast) {
```
Comment (bmuschko) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: This basically already exposed the functionality for the task type. In the build script it could look as follows.

```
test {
    failFast = true
}
```

That's what I meant by "expose a property on the task type".

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/processors/MaxNParallelTestClassProcessor.java (hunk 1 above)
Comment (oehme) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/processors/MaxNParallelTestClassProcessor.java]: This is a case of "leaky abstraction" and will probably give us some more issues when we poke at this with integration tests. If this works out I'm okay with doing it like that as a first step, but we should probably revisit the actor infrastructure.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    protected static final int MAX_WORKERS = 2
+
+    @Rule
+    BlockingHttpServer server = new BlockingHttpServer()
+
+    def setup() {
+        server.start()
+    }
+
+    def "all tests run without fail fast"() {
```
Comment (blindpirate) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy]: Since all these tests have similar structure, maybe you'd like to try [Data driven test](http://spockframework.org/spock/docs/1.1/data_driven_testing.html) with `where` clause. You can search `"where:"` in Gradle code base to see examples.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+                @Test
+                public void passingTest() {
+                    ${server.callFromBuild("$OTHER_RESOURCE")}
+                }
+            }
+        """.stripIndent()
+    }
+
+    private List<String> withNonfailingTests(int num) {
+        (1..num).collect {
```
Comment (blindpirate) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy]: I think [findResults](http://docs.groovy-lang.org/latest/html/groovy-jdk/java/lang/Iterable.html#findResults(groovy.lang.Closure)) is what you need here.

**Comment by lacasseio** [comment]
It seems to be missing test coverage for XCTest test framework. You should consider adding a test in `AbstractTestFrameworkIntegrationTest` to cover XCTest as well as any additional test framework that is going to be added to Gradle.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy (hunk 4 above)
Comment (acourtneybrown) [reply subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy]: the modulo is no longer necessary with the addition of `BlockingHttpServer.expectOptionalAndBlock()`

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy (hunk 3 above)
Comment (acourtneybrown) [reply subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy]: Based on your comment, I started heading down this path.  Once I got to the test checking the interaction between `forkEvery` and `maxParallelForks`, the data table was getting a bit unwieldy.  In addition, with the recent commits, `TestNGFailFastIntegrationTest` now has a test verifying the interaction with `parallel` in TestNG.

If you feel strongly, let me know & I'll make another attempt to refactor these test methods to use `where:` clauses.

**Comment by acourtneybrown** [comment]
@lacasseio, you are correct.  I had initially spiked on making the XCTest framework fail fast.  Now that the integration tests for JVM testing frameworks are written, I'm revisiting the XCTest.

**Comment by acourtneybrown** [comment]
@oehme @bmuschko @blindpirate - This PR should be ready for another round.

File: subprojects/docs/src/docs/userguide/javaPlugin.adoc [hunk 5]
```diff
@@ -563,6 +563,8 @@ The task has an `ignoreFailures` property to control the behavior when tests fai
 
 The `testLogging` property allows you to configure which test events are going to be logged and at which detail level. By default, a concise message will be logged for every failed test. See api:org.gradle.api.tasks.testing.logging.TestLoggingContainer[] for how to tune test logging to your preferences.
 
+The `failFast` property allows you to cause the `test` task to fail after the first test failure instead of running all tests. When this property is true, the resulting output will only show the results of tests that have completed up to and including the failure. The command line options `--fail-fast` and `--no-fail-fast` override the value of this property in the build file.  The default value for the `failFast` property is false.
```
Comment (bmuschko) [inline subprojects/docs/src/docs/userguide/javaPlugin.adoc]: Is it correct to assume that the use case for having `--no-fail-fast` is that you configured the `Test` task to fail fast but you still want to override behavior from the command line?

Comment (bmuschko) [inline subprojects/docs/src/docs/userguide/javaPlugin.adoc]: I think it would be good to directly link to the API at the beginning of the sentence where you mention the property.

Comment (bmuschko) [inline subprojects/docs/src/docs/userguide/javaPlugin.adoc]: It would be helpful to describe the use case here first. Why would a user want to use the feature? Something like "Projects with large test suite take a long time to execute even though a failure occurred early on causing unnecessary wait times (especially on CI). To short circuit the behavior...".

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 6]
```diff
@@ -544,6 +551,23 @@ public AbstractTestTask setTestNameIncludePatterns(List<String> testNamePattern)
         return this;
     }
 
+    @Option(option = "fail-fast", description = "Stops test execution after the first failed test")
+    public AbstractTestTask setFailFast(boolean ignored) {
```
Comment (bmuschko) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: The ignored parameters are somewhat unintuitive when reading the methods but I guess there's no way around it as you want to have `--no-fail-fast` as well. Some comment above: I am wondering if we actually need `--no-fail-fast`.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 7]
```diff
@@ -544,6 +551,23 @@ public AbstractTestTask setTestNameIncludePatterns(List<String> testNamePattern)
         return this;
     }
 
+    @Option(option = "fail-fast", description = "Stops test execution after the first failed test")
```
Comment (bmuschko) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: Using the description has a dot at the end of the sentence (see other usages of `@Option`).

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 8]
```diff
@@ -544,6 +551,23 @@ public AbstractTestTask setTestNameIncludePatterns(List<String> testNamePattern)
...
+
+    @Option(option = "no-fail-fast", description = "Ensures test execution continues after a failed test")
+    public AbstractTestTask setFailFastOff(boolean ignored) {
+        this.failFast = false;
+        return this;
+    }
+
+    @Input
+    public boolean getFailFast() {
+        return this.failFast;
```
Comment (bmuschko) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: The `this` is obsolete here.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+        withNonfailingTest()
+        def testExecution = server.expectConcurrentAndBlock(2, FAILED_RESOURCE, OTHER_RESOURCE)
+
+        when:
+        def gradleHandle = executer.withTasks('test', '--no-fail-fast').start()
+        testExecution.waitForAllPendingCalls()
+
+        then:
+        testExecution.release(FAILED_RESOURCE)
+        sleep(1000)
```
Comment (bmuschko) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy]: Why do we need the `sleep` here?

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+        testExecution.release(OTHER_RESOURCE)
+        gradleHandle.waitForFailure()
+        def result = new DefaultTestExecutionResult(testDirectory)
+        result.testClass('pkg.FailedTest').assertTestFailed('failTest', Matchers.anything())
+        result.testClass('pkg.OtherTest').assertTestPassed('passingTest')
+    }
+
+    def "stop test execution with failFast"() {
```
Comment (bmuschko) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy]: The tests that use the task property and the one using the command line flag look almost identical. I am sure we can consolidate them with a `where` clause.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+        testExecution.waitForAllPendingCalls()
+
+        then:
+        testExecution.release(FAILED_RESOURCE)
+        gradleHandle.waitForFailure()
+        assert gradleHandle.standardOutput.matches(/(?s).*pkg\.FailedTest.*failTest.*FAILED.*java.lang.RuntimeException at FailedTest.java.*/)
+        assert !gradleHandle.standardOutput.contains('pkg.OtherTest')
+    }
+
+    void withBuildFile(int maxWorkers = MAX_WORKERS, int forkEvery = 0) {
```
Comment (bmuschko) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy]: Might be worth to move those fixtures into a separate class. Add a constructor to the new class for passing in the `TestFile` instance so that you can create new `file` references.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 7 above)
Comment (acourtneybrown) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: done

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 6 above)
Comment (acourtneybrown) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: done

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 8 above)
Comment (acourtneybrown) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: done

File: subprojects/docs/src/docs/userguide/javaPlugin.adoc (hunk 5 above)
Comment (acourtneybrown) [reply subprojects/docs/src/docs/userguide/javaPlugin.adoc]: done

Comment (acourtneybrown) [reply subprojects/docs/src/docs/userguide/javaPlugin.adoc]: Yeah.  Updated the doc here to call that out.

Comment (acourtneybrown) [reply subprojects/docs/src/docs/userguide/javaPlugin.adoc]: done

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy (hunk 10 above)
Comment (acourtneybrown) [reply subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy]: done

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy (hunk 9 above)
Comment (acourtneybrown) [reply subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy]: They are an artifact of the early testing passes before I added `BlockingHttpServer.expectOptionalAndBlock()`, but are no longer necessary (& were never a good idea :) ).

Removed.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy (hunk 11 above)
Comment (acourtneybrown) [reply subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/fixture/AbstractJvmFailFastIntegrationSpec.groovy]: done (but it took a little more than just the `TestFile`)

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/FailFastTestListenerInternal.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+            if (testDescriptor.isComposite()) {
+                delegateResult = new DefaultTestResult(TestResult.ResultType.FAILURE, testResult.getStartTime(), testResult.getEndTime(), testResult.getTestCount(), testResult.getSuccessfulTestCount(), testResult.getFailedTestCount(), testResult.getExceptions());
+            } else {
+                delegateResult = new DefaultTestResult(TestResult.ResultType.SKIPPED, testResult.getStartTime(), testResult.getEndTime(), testResult.getTestCount(), testResult.getSuccessfulTestCount(), testResult.getFailedTestCount(), testResult.getExceptions());
+            }
+        }
+
+        delegate.completed(testDescriptor, delegateResult, completeEvent);
+
+        if (testResult.getResultType() == TestResult.ResultType.FAILURE) {
```
Comment (oehme) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/FailFastTestListenerInternal.java]: ⭕️ You might want to add `&& !failed` to avoid repeatedly telling the delegate to stop.

File: subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessorTest.groovy [hunk 13]
```diff
@@ -80,6 +81,63 @@ class ForkingTestClassProcessorTest extends Specification {
...
+        _ * workerProcessBuilder.getJavaCommand() >> Stub (JavaExecHandleBuilder)
+        1 * workerProcess.getConnection() >> Stub(ObjectConnection) { addOutgoing(_) >> remoteProcessor }
+
+        when:
+        processor.processTestClass(testClass)
+        processor.stopNow()
+        processor.stop()
+
+        then:
+        1 * workerLease.startChild() >> completion
```
Comment (oehme) [inline subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessorTest.groovy]: ❌ This applies to some of the other tests as well: This is over-asserting. Only mock the interactions that you are actually trying to test, ideally one per test. Otherwise you'll bind the test so tightly to the implementation that it'll fail for all kinds of unrelated reasons, making refactoring almost impossible.

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java [hunk 14]
```diff
@@ -424,6 +427,45 @@ public void setDebug(boolean enabled) {
         forkOptions.setDebug(enabled);
     }
 
+    /**
+     * Enables fail fast behavior causing the task will fail on the first failed test.
+     */
+    @Option(option = "fail-fast", description = "Stops test execution after the first failed test.")
+    public Test failFastOn(boolean ignored) {
```
Comment (oehme) [inline subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: ❌ Unnecessary additional methods. We only need the getter and setter.

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java [hunk 15]
```diff
@@ -424,6 +427,45 @@ public void setDebug(boolean enabled) {
...
+        setFailFast(true);
+        return this;
+    }
+
+    /**
+     * Disables fail fast behavior causing the task to run all tests.
+     */
+    @Option(option = "no-fail-fast", description = "Ensures test execution continues after a failed test.")
```
Comment (oehme) [inline subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: ❌ Not necessary, the user can use `--fail-fast=false`. If we think that `--no-fail-fast` is important, we should bake this as a general feature into our `@Option` handling, not add additional methods here.

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java [hunk 16]
```diff
@@ -424,6 +427,45 @@ public void setDebug(boolean enabled) {
...
+    public void setFailFast(boolean failFast) {
+        super.setFailFast(failFast);
+    }
+
+    /**
+     * Indicates if this task will fail on the first failed test
+     *
+     * @return whether this task will fail on the first failed test
+     */
+    @Input
```
Comment (oehme) [inline subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: ❌ This is not an input - it doesn't change the output of a successful execution. Failures are not cached by Gradle.

File: subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/xctest/internal/execution/XCTestExecuter.java [hunk 17]
```diff
@@ -55,6 +55,9 @@
  * - Test probing (so we know which tests exist without executing them)
  */
 public class XCTestExecuter implements TestExecuter<XCTestTestExecutionSpec> {
+
+    private TestClassProcessor processor;
```
Comment (oehme) [inline subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/xctest/internal/execution/XCTestExecuter.java]: ⭕️ Let's revert these changes for now.

File: subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/FailFastTestListenerTestInternal.groovy [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+package org.gradle.api.internal.tasks.testing
+
+import org.gradle.api.internal.tasks.testing.results.TestListenerInternal
+import org.gradle.api.tasks.testing.TestOutputEvent
+import org.gradle.api.tasks.testing.TestResult
+import spock.lang.Specification
+import spock.lang.Subject
+import spock.lang.Unroll
+
+class FailFastTestListenerTestInternal extends Specification {
```
Comment (oehme) [inline subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/FailFastTestListenerTestInternal.groovy]: ❌  This test is using mocking so much that it essentially repeats the implementation line by line. Only mock interactions that you care about. Use Stubs (or the real classes in case of simple data objects) for everything else.

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java (hunk 16 above)
Comment (acourtneybrown) [reply subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: done

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/FailFastTestListenerInternal.java (hunk 12 above)
Comment (acourtneybrown) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/FailFastTestListenerInternal.java]: done

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java (hunk 15 above)
Comment (acourtneybrown) [reply subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: @bmuschko had a similar comment, but it looks like `@Option` does not support this.  When I tried running with `--fail-fast=false`, Gradle outputs:
```
$ gradle test --fail-fast=false
...
FAILURE: Build failed with an exception.

* What went wrong:
Problem configuring task :test from command line.
> Command-line option '--fail-fast' does not take an argument.
```

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java (hunk 14 above)
Comment (acourtneybrown) [reply subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: see reply to next comment

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java (hunk 15 above)
Comment (bmuschko) [reply subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: @acourtneybrown I don't think the boolean option supports the behavior. You can only use `--fail-fast` (== true) or not use it at all (== false). See https://docs.gradle.org/nightly/userguide/custom_tasks.html#sec:supported_task_option_data_types.

Comment (acourtneybrown) [reply subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: @bmuschko @oehme after digging a little more, it looks like `AbstractOptionElement.calculateOptionType()` is where the code explicitly disallows passing `true` or `false` values to `[bB]oolean` command line args:

```
    protected static Class<?> calculateOptionType(Class<?> type) {
        //we don't want to support "--flag true" syntax
        if (type == Boolean.class || type == Boolean.TYPE) {
            return Void.TYPE;
        } else {
            return type;
        }
    }
```

Are you okay with having the 2 command line options (`--fail-fast` & `--no-fail-fast`) being handled explicitly?  Or do we want to add handling to `@Option` to automatically create a `no` version of `[bB]oolean` command line args?

File: subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/FailFastTestListenerTestInternal.groovy (hunk 18 above)
Comment (acourtneybrown) [reply subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/FailFastTestListenerTestInternal.groovy]: I've reworked this class to only mock out the `TestExecuter` and `TestListenerInternal` and used a bunch of the `Simple` implementation of the other (previously mocked) interfaces.  Let me know if the changes aren't what you were looking for.

File: subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/xctest/internal/execution/XCTestExecuter.java (hunk 17 above)
Comment (acourtneybrown) [reply subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/xctest/internal/execution/XCTestExecuter.java]: done, but the `stopNow` methods still need to exist for the interface, so they now throw `UnsupportedOperationException`s.

File: subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessorTest.groovy (hunk 13 above)
Comment (acourtneybrown) [reply subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessorTest.groovy]: I'm at a bit of a loss here.  Most of the mock interactions here are needed because they are returning objects(/mocks) that are used by the `ForkingTestClassProcessor`, either in managing its internal state or coordinating the underlying worker process.  Any concrete suggestions?

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java (hunk 15 above)
Comment (bmuschko) [reply subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: I don't think we should change the behavior of `@Option` right now. Plugin developers were already using the internal API of `@Option`. We also just made `@Option` a public API with 4.6.

In general I am wondering how often users will want to use the `false` option. Probably almost never. A user made the conscious decision to set `fail-fast` to `true` (likely by configuring the task). I cannot think of too many reasons why you'd want to temporary revert that decision. And if a user wants to not use fail-fast (even temporarily) then you can just change the flag in the build script or make it configurable via project property or something else.

Bottom line: My vote goes to just exposing the command line option `--fail-fast` and nothing else.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/BlockingHttpServer.java [hunk 19]
```diff
@@ -253,6 +253,18 @@ public BlockingHandler expectConcurrentAndBlock(int concurrent, String... expect
         return addBlockingHandler(concurrent, expectations);
     }
 
+    /**
+     * Expects exactly the given number of calls to be made concurrently from any combination of the optionally expected calls. Blocks each call until they are explicitly released.
+     * Since the expectations ore optional, they are still considered "complete" even if not all expected calls have been received.
```
Comment (blindpirate) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/http/BlockingHttpServer.java]: `ore` -> `are`

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java (hunk 15 above)
Comment (blindpirate) [reply subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: Let's spit this into 2 issues:

- Do we need the **feature** `--no-fail-fast` (or `--fail-fast false`, the name is not important)
- How do we implement it, by adding `--no-fail-fast` manually, or try something automatically?

For issue 2, I agree with @oehme , this should be generated automatically, for example, when `--fail-fast` is declared, `--no-fail-fast` is implicitly declared, we don't need to declare it manually. If so, issue1 seems to be solved as well - we don't need to declare it.

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/FailFastTestListenerInternal.java [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    public void completed(TestDescriptorInternal testDescriptor, TestResult testResult, TestCompleteEvent completeEvent) {
+        TestResult delegateResult = testResult;
+        if (failed) {
+            if (testDescriptor.isComposite()) {
+                delegateResult = new DefaultTestResult(TestResult.ResultType.FAILURE, testResult.getStartTime(), testResult.getEndTime(), testResult.getTestCount(), testResult.getSuccessfulTestCount(), testResult.getFailedTestCount(), testResult.getExceptions());
+            } else {
+                delegateResult = new DefaultTestResult(TestResult.ResultType.SKIPPED, testResult.getStartTime(), testResult.getEndTime(), testResult.getTestCount(), testResult.getSuccessfulTestCount(), testResult.getFailedTestCount(), testResult.getExceptions());
```
Comment (blindpirate) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/FailFastTestListenerInternal.java]: I'm not sure if there's potential risk or not. In my understanding, all test failure events after the initial failure will be marked as `SKIPPED`, I can understand why you're doing like this, but can't we keep it as a failure? This might result in more than one failed tests, but I feel ok with such result. 

Also, although this test result is marked as `SKIPPED`, `testResult.getTestCount()`, `testResult.getSuccessfulTestCount()`, `testResult.getFailedTestCount()` are still original value. This might be incorrect in the future statistic process (if there is).

**Comment by blindpirate** [comment]
Awesome work @acourtneybrown ! This looks good to me. You can trigger `Branch Build Accept` on our CI to validate your tests: https://builds.gradle.org/project.html?projectId=Gradle_Check&branch_Gradle_Check=adamb%2FstopNow-on-failed-test

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java (hunk 15 above)
Comment (oehme) [reply subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: @acourtneybrown I should have been clearer in my comment above: I think we should do *neither* in this PR. No extra option (since no other task does that) and no change to Option handling. 

We can revisit this later if anyone actually asks for it.

File: subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessorTest.groovy (hunk 13 above)
Comment (oehme) [reply subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessorTest.groovy]: Right now this test is asserting:

- how many modules a worker process loads
- how the worker gets its system properties
- that the worker will request a worker lease
- that the worker comes from a builder that comes from a factory

All of these are things that this test is not about. The majority of these interactions could be replaced by a `Stub` instead of a `Mock`. Don't let the current structure of the test class pin you down in that regard. E.g. you could replace most of fields at the top with `Stub`s and tighten the assertions in the tests where they really matter.

Comment (acourtneybrown) [reply subprojects/testing-base/src/test/groovy/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessorTest.groovy]: Makes more sense now.  I've reworked the test methods (in particular the ones around `stopNow`) to use `Stub()`s more & far fewer assertions on `Mock()`s.

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/FailFastTestListenerInternal.java (hunk 20 above)
Comment (acourtneybrown) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/FailFastTestListenerInternal.java]: @blindpirate Are you proposing that we only switch the `ResultType` to `SKIPPED` if the original result was `SUCCESS` (ie: retain existing `SKIPPED` and `FAILURE` results)?  If so, that seems reasonable.

As for the statistics, I'll have to look and see if they are incorporated when a test result is `SKIPPED` (unless @oehme knows off the top of his head).

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessor.java [hunk 21]
```diff
@@ -45,11 +47,13 @@
     private final Iterable<File> classPath;
     private final Action<WorkerProcessBuilder> buildConfigAction;
     private final ModuleRegistry moduleRegistry;
+    private final Lock lock = new ReentrantLock();
     private RemoteTestClassProcessor remoteProcessor;
     private WorkerProcess workerProcess;
     private TestResultProcessor resultProcessor;
     private WorkerLeaseRegistry.WorkerLeaseCompletion completion;
     private DocumentationRegistry documentationRegistry;
+    private volatile boolean stoppedNow;
```
Comment (oehme) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessor.java]: ⭕️ Doesn't need to be volatile if you're synchronizing access.

**Comment by oehme** [comment]
The failure on CI is a flaky test that was fixed on master. Rebasing should fix that. You might also want to squash your commits into one to make it easier for others to understand once merged.

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessor.java (hunk 21 above)
Comment (acourtneybrown) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessor.java]: Done ... the same thing occurred to me last night after I had signed off for the day.  I was planning on making this change first thing this morning.

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java (hunk 15 above)
Comment (bmuschko) [reply subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: How I understand @oehme's comment here is that we don't want _any_ command line option at the moment. None of the other Test configuration properties are exposed via the command line.

File: subprojects/docs/src/docs/userguide/javaPlugin.adoc [hunk 22]
```diff
@@ -563,6 +563,8 @@ The task has an `ignoreFailures` property to control the behavior when tests fai
 
 The `testLogging` property allows you to configure which test events are going to be logged and at which detail level. By default, a concise message will be logged for every failed test. See api:org.gradle.api.tasks.testing.logging.TestLoggingContainer[] for how to tune test logging to your preferences.
 
+Projects with large test suite can take a long time to execute even though a failure occurred early on leading to unnecessary wait times (especially on CI). To short circuit this behavior, the api:org.gradle.api.tasks.testing.Test#getFailFast()[`failFast`] property allows you to cause the `test` task to fail after the first test failure instead of running all tests. When this property is true, the resulting output will only show the results of tests that have completed up to and including the failure. The `--fail-fast` command line option enables the behavior from the command line.  Conversely, the '--no-fail-fast' command line option disables the behavior; this option is useful if the build file enables `failFast`, but you need to override it on a temporary basis. The default value for the `failFast` property is false.
```
Comment (bmuschko) [inline subprojects/docs/src/docs/userguide/javaPlugin.adoc]: :x: This is outdated either way. `--no-fail-fast` doesn't exist anymore.

Comment (acourtneybrown) [reply subprojects/docs/src/docs/userguide/javaPlugin.adoc]: Done.  I thought I had taken this out when I killed it in the design doc.  Thanks catching this.

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java (hunk 15 above)
Comment (acourtneybrown) [reply subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: And my interpretation of "no extra option" was to remove `--no-fail-fast` but keep `--fail-fast`.

@oehme can you clarify which way you want it?

**Comment by acourtneybrown** [comment]
@bmuschko - I added a small section to the release notes.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 2 above)
Comment (kheerthi) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: When using @option to accept values from user for a plugin, currently if a boolean variable is declared as false plugin, it can be enabled by passing --variable. However, since the variable needs to be true in most case and needs to be disabled on demand. How can this operation be performed from a consuming project?
